python/ray/data/tests/test_stats.py

```diff
@@ -133,6 +135,76 @@ def test_dataset_stats_sort(ray_start_regular_shared):
     assert "sort_reduce" in stats, stats
 
 
+@patch("ray.data._internal.plan.logger")
+def test_dataset_stats_autolog(mock_logger, ray_start_regular_shared):
+    """Tests functionality to autolog stats after each stage execution,
+    enabled via the `DatasetContext.enable_auto_log_stats` parameter."""
+    context = DatasetContext.get_current()
+    context.enable_auto_log_stats = True
+    ds = ray.data.range(1000, parallelism=10)
+    # Check logged stats after each stage
+    ds = ds.map_batches(lambda x: x)
+    logger_args, logger_kwargs = mock_logger.info.call_args
+    assert (
+        canonicalize(logger_args[0])
+        == """Stage Z read: N/N blocks executed in T, N/N blocks split from parent
+* Remote wall time: T min, T max, T mean, T total
+* Remote cpu time: T min, T max, T mean, T total
+* Peak heap memory usage (MiB): N min, N max, N mean
+* Output num rows: N min, N max, N mean, N total
+* Output size bytes: N min, N max, N mean, N total
+* Tasks per node: N min, N max, N mean; N nodes used
+"""
+    )
+
+    ds = ds.map(lambda x: x)
+    logger_args, logger_kwargs = mock_logger.info.call_args
+    assert (
+        canonicalize(logger_args[0])
+        == """Stage N read->map_batches: N/N blocks executed in T
+* Remote wall time: T min, T max, T mean, T total
+* Remote cpu time: T min, T max, T mean, T total
+* Peak heap memory usage (MiB): N min, N max, N mean
+* Output num rows: N min, N max, N mean, N total
+* Output size bytes: N min, N max, N mean, N total
+* Tasks per node: N min, N max, N mean; N nodes used
+"""
+    )
+
+    # Ensure that manual `stats()` method returns summarized stats
+    for batch in ds.iter_batches():
+        pass
+
+    stats = canonicalize(ds.stats())
+    assert (
+        stats
+        == """Stage N read->map_batches: N/N blocks executed in T
+* Remote wall time: T min, T max, T mean, T total
+* Remote cpu time: T min, T max, T mean, T total
+* Peak heap memory usage (MiB): N min, N max, N mean
+* Output num rows: N min, N max, N mean, N total
+* Output size bytes: N min, N max, N mean, N total
+* Tasks per node: N min, N max, N mean; N nodes used
+
+Stage N map: N/N blocks executed in T
+* Remote wall time: T min, T max, T mean, T total
+* Remote cpu time: T min, T max, T mean, T total
+* Peak heap memory usage (MiB): N min, N max, N mean
+* Output num rows: N min, N max, N mean, N total
+* Output size bytes: N min, N max, N mean, N total
+* Tasks per node: N min, N max, N mean; N nodes used
+
+Dataset iterator time breakdown:
+* In ray.wait(): T
+* In ray.get(): T
+* In next_batch(): T
+* In format_batch(): T
+* In user code: T
+* Total time: T
+"""
+    )
+
+
```

Comment:
Let's also add a test for `DatasetPipeline`. Checkout existing test in this file `test_dataset_pipeline_stats_basic`.